File: app/src/main/java/org/metabrainz/mobile/data/sources/api/OAuthAuthenticator.java
package org.metabrainz.mobile.data.sources.api;

import android.content.SharedPreferences;

import androidx.annotation.Nullable;

import org.metabrainz.mobile.data.sources.api.entities.AccessToken;
import org.metabrainz.mobile.presentation.features.login.LoginSharedPreferences;
import org.metabrainz.mobile.util.Log;

import java.io.IOException;

import okhttp3.Authenticator;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.Route;
import retrofit2.Call;

public class OAuthAuthenticator implements Authenticator {

    @Nullable
    @Override
    public Request authenticate(@Nullable Route route, Response response) throws IOException {

        LoginService service = MusicBrainzServiceGenerator.createService(LoginService.class,
                false);
        Log.d("OkHttp : " + response.body().string());

        String refreshToken = LoginSharedPreferences.getRefreshToken();

        Call<AccessToken> call = service.refreshAccessToken(refreshToken,
                "refresh_token",
                MusicBrainzServiceGenerator.CLIENT_ID,
                MusicBrainzServiceGenerator.CLIENT_SECRET);

        retrofit2.Response<AccessToken> newResponse = call.execute();

        AccessToken token = newResponse.body();
        if (token != null) {
            SharedPreferences.Editor editor = LoginSharedPreferences.getPreferences().edit();
            editor.putString(LoginSharedPreferences.REFRESH_TOKEN, token.getRefreshToken());
            editor.putString(LoginSharedPreferences.ACCESS_TOKEN, token.getAccessToken());
            editor.apply();
        }

        if (token != null && token.getAccessToken() != null)
            return response.request()
                    .newBuilder()
                    .header("Authorization", "Bearer " + token.getAccessToken())
                    .build();
        else return null;
    }
}

File: app/src/main/java/org/metabrainz/mobile/data/sources/api/OAuthAuthenticator.kt
package org.metabrainz.mobile.data.sources.api

import okhttp3.Authenticator
import okhttp3.Request
import okhttp3.Response
import okhttp3.Route
import org.metabrainz.mobile.data.sources.api.MusicBrainzServiceGenerator.createService
import org.metabrainz.mobile.presentation.features.login.LoginSharedPreferences
import org.metabrainz.mobile.util.Log.d
import java.io.IOException
import java.util.*

internal class OAuthAuthenticator : Authenticator {
    @Throws(IOException::class)
    override fun authenticate(route: Route?, response: Response): Request? {
        val service = createService(LoginService::class.java,
                false)
        d("OkHttp : " + Objects.requireNonNull(response.body)!!.string())
        val refreshToken = LoginSharedPreferences.refreshToken
        val call = service.refreshAccessToken(refreshToken,
                "refresh_token",
                MusicBrainzServiceGenerator.CLIENT_ID,
                MusicBrainzServiceGenerator.CLIENT_SECRET)
        val newResponse = call!!.execute()
        val token = newResponse.body()
        if (token != null) {
            val editor = LoginSharedPreferences.preferences.edit()
            editor.putString(LoginSharedPreferences.REFRESH_TOKEN, token.refreshToken)
            editor.putString(LoginSharedPreferences.ACCESS_TOKEN, token.accessToken)
            editor.apply()
        }
        return if (token?.accessToken != null) response.request
                .newBuilder()
                .header("Authorization", "Bearer " + token.accessToken)
                .build() else null
    }
}